lib/service/service.go

```diff
@@ -2339,52 +2335,31 @@ func (process *TeleportProcess) initSSH() error {
 				return trace.Wrap(err)
 			}
 			log.Infof("Service is starting in tunnel mode.")
-
-			// Broadcast that the node has started.
-			process.BroadcastEvent(Event{Name: NodeSSHReady, Payload: nil})
 		}
 
+		// Broadcast that the node has started.
+		process.BroadcastEvent(Event{Name: NodeSSHReady, Payload: nil})
+
 		// Block and wait while the node is running.
 		s.Wait()
-		if conn.UseTunnel() {
-			agentPool.Wait()
-		}
+		agentPool.Wait()
 
-		log.Infof("Exited.")
-		return nil
-	})
+		event, err := process.WaitForEvent(process.ExitContext(), TeleportExitEvent)
```

Comment:
The `s.Wait()` and `agentPool.Wait()` calls should go after the `WaitForEvent` and the close/shutdown, even though technically `s.Wait()` can return on a network error on the listener, but then we'd still get stuck on `agentPool.Wait()` anyway, which will never exit until the agent pool is explicitly stopped.